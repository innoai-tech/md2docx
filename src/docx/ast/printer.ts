import { isBuffer, isNull, isUndefined, map, repeat } from "lodash";
import type { FileChild } from "./nodes.ts";

export const printAST = (children: FileChild[]): string => {
  const print = (root: any, depth = 0) => {
    const { children: _1, type: _2, data: _3, options } = root as any;

    const attrsStr = map(options, (v, k) => {
      if (isNull(v) || isUndefined(v)) {
        return "";
      }
      if (isBuffer(v)) {
        return `${k}=[Buffer]`;
      }

      return `${k}=${JSON.stringify(v)}`;
    })
      .filter((v) => v)
      .sort()
      .join(",");

    const childrenStr: string =
      root.children
        ?.map((c: any) => {
          return `${repeat("  ", depth + 1)}${print(c as any, depth + 1)}`;
        })
        .join("\n") ?? "";

    return `${root.type}${attrsStr ? `{${attrsStr}}` : ""}${
      childrenStr ? `\n${childrenStr}` : ""
    }`;
  };

  return children.map((c) => print(c)).join("\n");
};
