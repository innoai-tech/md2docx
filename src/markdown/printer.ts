import { type Root } from "mdast";
import { isBuffer, isNull, isUndefined, map, repeat } from "lodash";

export const printAST = (root: Root): string => {
  const print = (root: Root, depth = 0) => {
    const {
      position: _0,
      children: _1,
      type: _3,
      data,
      attributes,
      ...props
    } = root as any;

    const attrsStr = map({ ...data, ...props, ...attributes }, (v, k) => {
      if (isNull(v) || isUndefined(v)) {
        return "";
      }
      if (isBuffer(v)) {
        return `${k}=[Buffer]`;
      }

      return `${k}=${JSON.stringify(v)}`;
    })
      .filter((v) => v)
      .sort()
      .join(",");

    const childrenStr: string =
      root.children
        ?.map((c) => {
          return `${repeat("  ", depth + 1)}${print(c as any, depth + 1)}`;
        })
        .join("\n") ?? "";

    return `${root.type}${attrsStr ? `{${attrsStr}}` : ""}${
      childrenStr ? `\n${childrenStr}` : ""
    }`;
  };

  return print(root);
};
