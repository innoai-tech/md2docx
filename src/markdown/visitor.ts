import type {
  Root,
  DefinitionContent,
  PhrasingContent,
  BlockContent,
  RootContent,
} from "mdast";
import type { VFile } from "vfile";
import { parse } from "./markdown.ts";
import { Cursor } from "./Cursor.ts";
import type { FS } from "../fs";
import { OsFS } from "../fs";

export type Exactly<Union, Type> = Union extends { type: Type } ? Union : never;

export type MdastNode =
  | RootContent
  | BlockContent
  | PhrasingContent
  | DefinitionContent;

export type Visitors = {
  [K in MdastNode["type"]]?: (
    c: Cursor<Exactly<MdastNode, K>>,
    ctx: VisitorContext,
  ) => Promise<void>;
};

export type VisitorContext = {
  fs: FS;
  parser: Parser;
  visitChildren(c: Cursor, options?: ProcessOptions): Promise<void>;
};

export type ProcessOptions = {
  fs?: FS;
  visitors?: Visitors[];
};

export interface Parser {
  parse: (f: VFile) => Cursor<Root>;
}

export const process = async (
  file: VFile,
  options: ProcessOptions,
): Promise<Root> => {
  const w = new Context(
    {
      parse: (file) => {
        return Cursor.of(parse(file.value), { file });
      },
    },
    options.fs ?? new OsFS(),
    options.visitors ?? [],
  );

  const root = w.parser.parse(file);
  await w.visitChildren(root);
  return root.node;
};

class Context implements VisitorContext {
  constructor(
    public parser: Parser,
    public fs: FS,
    protected visitors: Visitors[],
  ) {}

  async visitChildren(c: Cursor, options: ProcessOptions = {}): Promise<void> {
    for (const child of c.children()) {
      for (const v of [...this.visitors, ...(options.visitors ?? [])]) {
        const visitor = (v as any)[child.node.type];

        if (!visitor) {
          await this.visitChildren(child);
          continue;
        }

        await visitor(child, this);
      }
    }
  }
}
