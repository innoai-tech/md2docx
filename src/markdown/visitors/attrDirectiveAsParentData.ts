import { type Visitors } from "../visitor.ts";
import type { Directives } from "mdast-util-directive";

export const attrDirectiveAsParentData = (): Visitors => {
  const toData = (c: Directives) => {
    const { id, class: className, ...props } = c.attributes as any;

    return {
      ...props,
      class: className,
      id,
    };
  };

  return {
    async textDirective(c, _) {
      switch (c.node.name) {
        case "attr": {
          if (c.parent) {
            c.parent.patchData(toData(c.node));
            c.remove();
          }
        }
      }
    },
  };
};
