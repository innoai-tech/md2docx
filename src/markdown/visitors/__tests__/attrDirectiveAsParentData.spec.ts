import { describe, expect, it } from "bun:test";
import { VFile } from "vfile";
import { process } from "../../visitor.ts";
import { printAST } from "../../printer.ts";
import { attrDirectiveAsParentData } from "../attrDirectiveAsParentData.ts";

describe("#attrDirectiveAsParentData", () => {
  it("textDirective should patch parent data", async () => {
    const node = await process(
      new VFile(`
# 一级标题:attr{#h1}
`),
      {
        visitors: [attrDirectiveAsParentData()],
      },
    );

    expect(printAST(node)).toMatchSnapshot();
  });
});
