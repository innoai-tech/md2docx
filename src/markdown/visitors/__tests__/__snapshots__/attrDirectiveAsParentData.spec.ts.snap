// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`#attrDirectiveAsParentData textDirective should patch parent data 1`] = `
"root
  heading{depth=1,id="h1"}
    text{value="一级标题"}"
`;
